ldcpy/metrics.py

```diff
@@ -1058,83 +1150,125 @@ def ssim_value_fp(self):
+                    a1a2_cov = np.average(
+                        (a1_win[indices1] - a1_mu) * (a2_win[indices2] - a2_mu),
+                        weights=Wt[indices1],
+                    )
 
                     # SSIM for this window
                     # first term
                     ssim_t1 = 2 * a1_mu * a2_mu
                     ssim_b1 = a1_mu * a1_mu + a2_mu * a2_mu
-                    # print("term1 t/b = ", ssim_t1, ssim_b1)
-                    if ssim_b1 < my_eps:
-                        C1 = my_eps
-                        # print("TERM 1 trigger for C1")
-                    else:
-                        C1 = 0.0
+                    C1 = my_eps
+
                     ssim_t1 = ssim_t1 + C1
                     ssim_b1 = ssim_b1 + C1
 
                     # second term
                     ssim_t2 = 2 * a1a2_cov
                     ssim_b2 = a1_std_sq + a2_std_sq
-                    # print("term2 t/b= ", ssim_t2, ssim_b2)
-                    if abs(ssim_b2) < my_eps:
-                        C2 = my_eps
-                        # print("TERM 2 trigger for C2")
-                        if ssim_t2 < my_eps:
-                            C3 = my_eps
-                            # print("TERM 2 trigger for C3")
-                        else:
-                            C3 = 0.0
-                    else:
-                        C2 = 0.0
-                        C3 = 0.0
+                    C2 = C3 = my_eps
+
                     ssim_t2 = ssim_t2 + C3
                     ssim_b2 = ssim_b2 + C2
 
                     ssim_1 = ssim_t1 / ssim_b1
                     ssim_2 = ssim_t2 / ssim_b2
                     ssim_mat[i, j] = ssim_1 * ssim_2
 
-            # mean_ssim = np.average(ssim_mat)
+            mean_ssim = np.nanmean(ssim_mat)
             # to ignore edge effects (as in skimage)
-            mean_ssim = np.average(crop(ssim_mat, k))
+            # mean_ssim = np.nanmean(crop(ssim_mat, k))
 
             self._ssim_value_fp = mean_ssim
 
         return self._ssim_value_fp
 
     @property
     def ssim_value_fp_old(self):
+        """To mimc what zchecker does - the ssim on the fp data with
+        original constants and no scaling. This will return Nan on POP data.
+        """
+
+        import numpy as np
+        from skimage.metrics import structural_similarity as ssim
+
+        if not self._is_memoized('_ssim_value_fp_old'):
+
+            # get 2D arrays
+            a1 = self._metrics1.get_metric('ds').data
+            a2 = self._metrics2.get_metric('ds').data
+
+            if dask.is_dask_collection(a1):
+                a1 = a1.compute()
+            if dask.is_dask_collection(a2):
+                a2 = a2.compute()
+
+            # if this is a 3D variable, we will just do level 0 for now...
+            # (consider doing each level seperately)
+            if self._metrics1._vert_dim_name is not None:
+                a1 = a1[0, :, :]
+                a2 = a2[0, :, :]
```

Comment:
Aren't there cases in which this assumption (that the vertical level is the first dimension) is incorrect? Would it be more robust if we were to keep the initial `a1` and `a2` as xarray objects and then just use `.isel()` or `.sel()`?

Reply:
good point - that is likely true - i'll change this

Reply:
@andersy005 

To fix this, I need to use a variable for isel/sel, but I can't figure out how to do that.  It seems that I can't do this:

  if self._metrics1._vert_dim_name is not None:
                vname = self._metrics1._vert_dim_name
                print(vname)
                a1 = self._metrics1.get_metric('ds').isel(vname=0).data

I get this message:
 lev
 ValueError: dimensions {'vname'} do not exist. Expected one or more of ('lev', 'lat', 'lon')

Reply:
@allibco, you want to use the dictionary notation here. 


Replace 

```python
a1 = self._metrics1.get_metric('ds').isel(vname=0).data
``` 

with 

```python
a1 = self._metrics1.get_metric('ds').isel({vname: 0}).data
```

Reply:
thanks - i didn't know about that